xdsl/dialects/x86/ops.py

```diff
@@ -276,6 +285,112 @@ class MovOp(RROperation[GeneralRegisterType, GeneralRegisterType]):
     name = "x86.mov"
 
 
+class ROperation(Generic[R1InvT], SingleOperandInstruction):
+    """
+    A base class for x86 operations that have one register.
+    """
+
+    source = opt_operand_def(R1InvT)
+    destination = opt_result_def(R1InvT)
```

Comment:
why optional? I'd expect these to be required

Reply:
It’s to support different types of operations eg. pop has only a destination while push has only a source.

Reply:
It sounds like they should be different operations, one with only a source, and one with only a destination. Otherwise a user may construct a pop with only a source.

Reply:
I made the verification methods to avoid that, but can split them into two methods as well if that’s better

Reply:
Yes please, verification is good, but declarative structures are better, that way we get more warnings statically rather than at runtime.

Reply:
What to do if the operand is both a source and a destination like in the 'not' instruction? a seperate class for this case as well?

Sorry for no progress by the way, i'll get on it after I'm done with the dissertation writing,

Reply:
Then I'd say we want an instruction that has a source and target register, and constrains the type of the operand and result to be exactly the same, with a ConstraintVar, something like this:
``` python
    T = Annotated[RegisterType, ConstraintVar("T")]
    operand: Operand = operand_def(T)
    result: OpResult = result_def(T)
```